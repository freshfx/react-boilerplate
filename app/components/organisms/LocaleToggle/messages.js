/*
 * LocaleToggle Messages
 *
 * This contains all the text for the LanguageToggle component.
 */
import {defineMessages} from 'react-intl'

const scope = 'boilerplate.organisms.LocaleToggle'

export default defineMessages({
  de: {
    defaultMessage: 'de',
    id: `${scope}.de`
  },
  en: {
    defaultMessage: 'en',
    id: `${scope}.en`
  }
})
