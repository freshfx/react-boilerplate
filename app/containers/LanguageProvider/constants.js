/*
 *
 * LanguageProvider constants
 *
 */

export const CHANGE_LOCALE = 'app/LanguageToggle/CHANGE_LOCALE'

export default {CHANGE_LOCALE}
