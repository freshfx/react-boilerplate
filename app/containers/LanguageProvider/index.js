/*
 *
 * LanguageProvider
 *
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import React from 'react'
import PropTypes from 'prop-types'
import {compose} from 'redux'
import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'
import {IntlProvider} from 'react-intl'

import {selectors} from 'modules/language'

export class LanguageProvider extends React.PureComponent {
  // eslint-disable-line react/prefer-stateless-function
  render() {
    return (
      <IntlProvider
        locale={this.props.locale}
        key={this.props.locale}
        messages={this.props.messages[this.props.locale]}
        textComponent="span"
      >
        {React.Children.only(this.props.children)}
      </IntlProvider>
    )
  }
}

LanguageProvider.propTypes = {
  children: PropTypes.element.isRequired,
  locale: PropTypes.string.isRequired,
  messages: PropTypes.object
}

LanguageProvider.defaultProps = {
  messages: {}
}

const mapStateToProps = createStructuredSelector({
  locale: selectors.selectLocale
})

const withConnect = connect(mapStateToProps)

export default compose(withConnect)(LanguageProvider)
