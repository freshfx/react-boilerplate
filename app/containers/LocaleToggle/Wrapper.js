import styled from '@emotion/styled'

const Wrapper = styled.div`
  padding: 2px;
`

export default Wrapper
