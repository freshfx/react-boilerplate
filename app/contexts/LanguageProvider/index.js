/*
 *
 * LanguageProvider
 *
 * this component connects the redux state language locale to the
 * IntlProvider component and i18n messages (loaded from `app/translations`)
 */

import React from 'react'
import PropTypes from 'prop-types'
import {compose} from 'redux'
import {useSelector} from 'react-redux'
import {IntlProvider} from 'react-intl'

import {selectors} from 'modules/language'

const LanguageProvider = ({children, messages}) => {
  const locale = useSelector(selectors.selectLocale)

  return (
    <IntlProvider
      locale={locale}
      key={locale}
      messages={messages[locale]}
      textComponent="span"
    >
      {React.Children.only(children)}
    </IntlProvider>
  )
}

LanguageProvider.propTypes = {
  children: PropTypes.element.isRequired,
  messages: PropTypes.object
}

LanguageProvider.defaultProps = {
  messages: {}
}

export default compose(React.memo)(LanguageProvider)
