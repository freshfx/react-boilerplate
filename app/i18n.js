/**
 * I18n.js
 *
 * This will setup the i18n language files and locale data for your app.
 *
 *   IMPORTANT: This file is used by the internal build
 *   script `extract-intl`, and must use CommonJS module syntax
 *   You CANNOT use import/export in this file.
 */
const enTranslationMessages = require('./translations/en.json')
const deTranslationMessages = require('./translations/de.json')

/* eslint-disable global-require */
/* istanbul ignore if */
if (!Intl.PluralRules) {
  require('intl-pluralrules/polyfill')
  require('@formatjs/intl-pluralrules/dist/locale-data/en')
  require('@formatjs/intl-pluralrules/dist/locale-data/de')
}

/* istanbul ignore if */
if (!Intl.RelativeTimeFormat) {
  require('@formatjs/intl-relativetimeformat/polyfill')
  require('@formatjs/intl-relativetimeformat/dist/locale-data/en')
  require('@formatjs/intl-relativetimeformat/dist/locale-data/de')
}

const DEFAULT_LOCALE = 'en'

// Prettier-ignore
const appLocales = [
  'en',
  'de'
]

const translations = {
  formatTranslationMessages: function formatTranslationMessages(locale, messages) {
    const defaultFormattedMessages = this.getDefaultFormattedMessages(locale)
    const flattenFormattedMessages = (formattedMessages, key) => {
      let value = messages[key]
      if (!messages[key] && locale !== DEFAULT_LOCALE) {
        value = defaultFormattedMessages[key]
      }
      return Object.assign(formattedMessages, {[key]: value})
    }
    return Object.keys(messages).reduce(flattenFormattedMessages, {})
  },

  getDefaultFormattedMessages: function getDefaultFormattedMessages(locale) {
    if (locale !== DEFAULT_LOCALE) {
      return this.formatTranslationMessages(DEFAULT_LOCALE, enTranslationMessages)
    }
    return {}
  }
}

const translationMessages = {
  de: translations.formatTranslationMessages('de', deTranslationMessages),
  en: translations.formatTranslationMessages('en', enTranslationMessages)
}

exports.appLocales = appLocales
exports.formatTranslationMessages = translations.formatTranslationMessages.bind(translations)
exports.translationMessages = translationMessages
exports.DEFAULT_LOCALE = DEFAULT_LOCALE
