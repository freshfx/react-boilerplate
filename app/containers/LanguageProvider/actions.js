/*
 *
 * LanguageProvider actions
 *
 */

import {CHANGE_LOCALE} from './constants'

export const changeLocale = languageLocale => ({
  locale: languageLocale,
  type: CHANGE_LOCALE
})

export default {changeLocale}
