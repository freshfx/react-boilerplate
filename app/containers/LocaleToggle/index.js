/*
 *
 * LanguageToggle
 *
 */

import React from 'react'
import {compose} from 'redux'
import {useDispatch, useSelector} from 'react-redux'

import {appLocales} from 'i18n'
import Toggle from 'components/Toggle'
import {actions, selectors} from 'modules/language'

import Wrapper from './Wrapper'
import messages from './messages'

const LocaleToggle = () => {
  const dispatch = useDispatch()
  const locale = useSelector(selectors.selectLocale)

  const onChangeLocale = ({target: {value}}) => {
    dispatch(actions.changeLocale({locale: value}))
  }

  return (
    <Wrapper>
      <Toggle
        value={locale}
        values={appLocales}
        messages={messages}
        onToggle={onChangeLocale}
      />
    </Wrapper>
  )
}

export default compose(React.memo)(LocaleToggle)
