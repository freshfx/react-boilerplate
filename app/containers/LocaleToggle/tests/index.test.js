import React from 'react'
import {Provider} from 'react-redux'
import {createMemoryHistory as createHistory} from 'history'
import {mount, shallow} from 'enzyme'

import {actions} from 'modules/language'

import LocaleToggle, {mapDispatchToProps} from '../index'
import LanguageProvider from '../../LanguageProvider'

import configureStore from '../../../configure-store'
import {translationMessages} from '../../../i18n'

const history = createHistory()

describe('<LocaleToggle />', () => {
  let store = {}

  beforeAll(() => {
    store = configureStore({}, history)
  })

  it('should render the default language messages', () => {
    const renderedComponent = shallow(<Provider store={store}>
      <LanguageProvider messages={translationMessages}>
        <LocaleToggle />
      </LanguageProvider>
    </Provider>)
    expect(renderedComponent.contains(<LocaleToggle />)).toBe(true)
  })

  it('should present the default `en` english language option', () => {
    const renderedComponent = mount(<Provider store={store}>
      <LanguageProvider messages={translationMessages}>
        <LocaleToggle />
      </LanguageProvider>
    </Provider>)
    expect(renderedComponent.contains(<option value="en">en</option>)).toBe(true)
  })

  describe('mapDispatchToProps', () => {
    describe('onLocaleToggle', () => {
      it('should be injected', () => {
        const dispatch = jest.fn()
        const result = mapDispatchToProps(dispatch)
        expect(result.onChangeLocale).toBeDefined()
      })

      it('should dispatch changeLocale when called', () => {
        const dispatch = jest.fn()
        const result = mapDispatchToProps(dispatch)
        const locale = 'de'
        result.onChangeLocale(locale)
        expect(dispatch).toHaveBeenCalledWith(actions.changeLocale({locale}))
      })
    })
  })
})
