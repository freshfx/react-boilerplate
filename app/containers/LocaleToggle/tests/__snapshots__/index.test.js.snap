// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleToggle should render the default language messages 1`] = `
.emotion-2 {
  padding: 2px;
}

.emotion-0 {
  line-height: 1em;
  background-color: transparent;
  border-style: none;
}

<div
  class="emotion-2 emotion-3"
>
  <select
    class="emotion-0 emotion-1"
  >
    <option
      value="en"
    >
      en
    </option>
    <option
      value="de"
    >
      de
    </option>
  </select>
</div>
`;
